<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>StabMap: Mosaic single cell data integration using non-overlapping features • StabMap</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="StabMap: Mosaic single cell data integration using non-overlapping features">
<meta property="og:description" content="StabMap">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">StabMap</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/stabMap_PBMC_Multiome.html">StabMap: Mosaic single cell data integration using non-overlapping features</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MarioniLab/StabMap/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>StabMap: Mosaic single cell data integration
using non-overlapping features</h1>
                        <h4 data-toc-skip class="author">Shila
Ghazanfar</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MarioniLab/StabMap/blob/HEAD/vignettes/stabMap_PBMC_Multiome.Rmd" class="external-link"><code>vignettes/stabMap_PBMC_Multiome.Rmd</code></a></small>
      <div class="hidden name"><code>stabMap_PBMC_Multiome.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marionilab.github.io/StabMap" class="external-link">StabMap</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellMultiModal</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2021</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>StabMap is a technique for performing mosaic single cell data
integration.</p>
<p>In this vignette we will elaborate on how these steps are implemented
in the <code>StabMap</code> package.</p>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellMultiModal/man/scMultiome.html" class="external-link">scMultiome</a></span><span class="op">(</span><span class="st">"pbmc_10x"</span>, mode <span class="op">=</span> <span class="st">"*"</span>, dry.run <span class="op">=</span> <span class="cn">FALSE</span>, format <span class="op">=</span> <span class="st">"MTX"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## snapshotDate(): 2021-10-19</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_atac_se.rds</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_atac.mtx.gz</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_rna_se.rds</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_rna.mtx.gz</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_atac,</span>
<span class="co">##  pbmc_rna</span></code></pre>
<pre><code><span class="co">## see ?SingleCellMultiModal and browseVignettes('SingleCellMultiModal') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<pre><code><span class="co">## see ?SingleCellMultiModal and browseVignettes('SingleCellMultiModal') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_atac,</span>
<span class="co">##  pbmc_rna</span></code></pre>
<pre><code><span class="co">## see ?SingleCellMultiModal and browseVignettes('SingleCellMultiModal') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<pre><code><span class="co">## see ?SingleCellMultiModal and browseVignettes('SingleCellMultiModal') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_colData</span></code></pre>
<pre><code><span class="co">## Working on: pbmc_sampleMap</span></code></pre>
<pre><code><span class="co">## see ?SingleCellMultiModal and browseVignettes('SingleCellMultiModal') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<pre><code><span class="co">## see ?SingleCellMultiModal and browseVignettes('SingleCellMultiModal') for documentation</span></code></pre>
<pre><code><span class="co">## loading from cache</span></code></pre>
<p>Perform some exploration of this data.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mae</span></code></pre></div>
<pre><code><span class="co">## A MultiAssayExperiment object of 2 listed</span>
<span class="co">##  experiments with user-defined names and respective classes.</span>
<span class="co">##  Containing an ExperimentList class object of length 2:</span>
<span class="co">##  [1] atac: SingleCellExperiment with 108344 rows and 10032 columns</span>
<span class="co">##  [2] rna: SingleCellExperiment with 36549 rows and 10032 columns</span>
<span class="co">## Functionality:</span>
<span class="co">##  experiments() - obtain the ExperimentList instance</span>
<span class="co">##  colData() - the primary/phenotype DataFrame</span>
<span class="co">##  sampleMap() - the sample coordination DataFrame</span>
<span class="co">##  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span>
<span class="co">##  *Format() - convert into a long or wide DataFrame</span>
<span class="co">##  assays() - convert ExperimentList to a SimpleList of matrices</span>
<span class="co">##  exportClass() - save data to flat files</span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/upsetSamples.html" class="external-link">upsetSamples</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## DataFrame with 6 rows and 6 columns</span>
<span class="co">##                  nCount_RNA nFeature_RNA nCount_ATAC nFeature_ATAC</span>
<span class="co">##                   &lt;integer&gt;    &lt;integer&gt;   &lt;integer&gt;     &lt;integer&gt;</span>
<span class="co">## AAACAGCCAAGGAATC       8380         3308       55582         13878</span>
<span class="co">## AAACAGCCAATCCCTT       3771         1896       20495          7253</span>
<span class="co">## AAACAGCCAATGCGCT       6876         2904       16674          6528</span>
<span class="co">## AAACAGCCAGTAGGTG       7614         3061       39454         11633</span>
<span class="co">## AAACAGCCAGTTTACG       3633         1691       20523          7245</span>
<span class="co">## AAACAGCCATCCAGGT       7782         3028       22412          8602</span>
<span class="co">##                                celltype broad_celltype</span>
<span class="co">##                             &lt;character&gt;    &lt;character&gt;</span>
<span class="co">## AAACAGCCAAGGAATC      naive CD4 T cells       Lymphoid</span>
<span class="co">## AAACAGCCAATCCCTT     memory CD4 T cells       Lymphoid</span>
<span class="co">## AAACAGCCAATGCGCT      naive CD4 T cells       Lymphoid</span>
<span class="co">## AAACAGCCAGTAGGTG      naive CD4 T cells       Lymphoid</span>
<span class="co">## AAACAGCCAGTTTACG     memory CD4 T cells       Lymphoid</span>
<span class="co">## AAACAGCCATCCAGGT non-classical monocy..        Myeloid</span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"rna"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 36549 10032</span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "atac" "rna"</span></code></pre>
<p>Normalise and select features for the RNA modality.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"rna"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># Normalisation</span>
<span class="va">sce.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span>

<span class="co"># Feature selection</span>
<span class="va">decomp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span>
<span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">decomp</span><span class="op">)</span><span class="op">[</span><span class="va">decomp</span><span class="op">$</span><span class="va">mean</span><span class="op">&gt;</span><span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">decomp</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 952</span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce.rna</span> <span class="op">&lt;-</span> <span class="va">sce.rna</span><span class="op">[</span><span class="va">hvgs</span>,<span class="op">]</span></code></pre></div>
<p>Normalise and select features for the ATAC modality.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"atac"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 108344  10032</span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://waldronlab.io/MultiAssayExperiment/reference/MultiAssayExperiment-methods.html" class="external-link">experiments</a></span><span class="op">(</span><span class="va">mae</span><span class="op">)</span><span class="op">[[</span><span class="st">"atac"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># Normalise</span>
<span class="va">sce.atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.atac</span><span class="op">)</span>

<span class="co"># Feature selection using highly variable peaks</span>
<span class="co"># And adding matching peaks to genes</span>
<span class="va">decomp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce.atac</span><span class="op">)</span>
<span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">decomp</span><span class="op">)</span><span class="op">[</span><span class="va">decomp</span><span class="op">$</span><span class="va">mean</span><span class="op">&gt;</span><span class="fl">0.25</span>
                         <span class="op">&amp;</span> <span class="va">decomp</span><span class="op">$</span><span class="va">p.value</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 788</span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce.atac</span> <span class="op">&lt;-</span> <span class="va">sce.atac</span><span class="op">[</span><span class="va">hvgs</span>,<span class="op">]</span></code></pre></div>
<p>Create a composite full data matrix by concatenating.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">logcounts_all</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce.atac</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1]  1740 10032</span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">assayType</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce.rna</span><span class="op">)</span>,
                   <span class="st">"rna"</span>, <span class="st">"atac"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">assayType</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## assayType</span>
<span class="co">## atac  rna </span>
<span class="co">##  788  952</span></code></pre>
</div>
<div class="section level2">
<h2 id="mosaic-data-integration-with-stabmap">Mosaic data integration with StabMap<a class="anchor" aria-label="anchor" href="#mosaic-data-integration-with-stabmap"></a>
</h2>
<p>We will simulate a situation where half of the cells correspond to
the Multiome modality, and half of the cells correspond to the RNA
modality. Our goal is to then generate a joint embedding of the cells
using all data, and to impute the missing ATAC values from the RNA
modality cells.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataType</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span>, <span class="st">"Multiome"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span>,
                           prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>,<span class="fl">0.5</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
                    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dataType</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## dataType</span>
<span class="co">## Multiome      RNA </span>
<span class="co">##     4983     5049</span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">assay_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  RNA <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span><span class="op">)</span>, <span class="va">dataType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">]</span>,
  Multiome <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"atac"</span><span class="op">)</span>, <span class="va">dataType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span><span class="op">]</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list</span>, <span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $RNA</span>
<span class="co">## [1]  952 5049</span>
<span class="co">## </span>
<span class="co">## $Multiome</span>
<span class="co">## [1] 1740 4983</span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list</span>, <span class="va">class</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $RNA</span>
<span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span>
<span class="co">## </span>
<span class="co">## $Multiome</span>
<span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span></code></pre>
<p>Examine the shared features between the two datasets using
<code><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet()</a></code>.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet</a></span><span class="op">(</span><span class="va">assay_list</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: UpSetR</span></code></pre>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>From this we note that there are shared features between the RNA and
Multiome datasets, but there are many features that are observed only in
the Multiome dataset and not the RNA - as we had constructed.</p>
<p>We can understand the <code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code> of these
datasets, which generates an <code>igraph</code> object, which can be
inspected and plotted.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mdt</span> <span class="op">=</span> <span class="fu"><a href="../reference/mosaicDataTopology.html">mosaicDataTopology</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">)</span>
<span class="va">mdt</span></code></pre></div>
<pre><code><span class="co">## IGRAPH 41ed2a4 UNW- 2 1 -- </span>
<span class="co">## + attr: name (v/c), frame.color (v/c), color (v/c), label.color (v/c),</span>
<span class="co">## | label.family (v/c), weight (e/n)</span>
<span class="co">## + edge from 41ed2a4 (vertex names):</span>
<span class="co">## [1] RNA--Multiome</span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mdt</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>From this we note that the datasets RNA and Multiome share at least
some features. StabMap requires that the mosaic data topology network be
connected, that is, that there should be a path between every pair of
nodes in the network.</p>
<p>We generate a common joint embedding for these data using StabMap.
Since the Multiome data contains all features, we treat this as the
reference dataset. Since we already examined the mosaic data topology,
we set <code>plot = FALSE</code>.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stab</span> <span class="op">=</span> <span class="fu"><a href="../reference/StabMap.html">stabMap</a></span><span class="op">(</span><span class="va">assay_list</span>,
               reference_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span>,
               plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: scater</span></code></pre>
<pre><code><span class="co">## Warning: package 'scater' was built under R version 4.1.1</span></code></pre>
<pre><code><span class="co">## Loading required package: ggplot2</span></code></pre>
<pre><code><span class="co">## Warning: package 'ggplot2' was built under R version 4.1.2</span></code></pre>
<pre><code><span class="co">## treating "Multiome" as reference</span></code></pre>
<pre><code><span class="co">## generating embedding for path with reference "Multiome": "Multiome"</span></code></pre>
<pre><code><span class="co">## generating embedding for path with reference "Multiome": "RNA" -&gt; "Multiome"</span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 10032    50</span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                  Multiome_PC1 Multiome_PC2 Multiome_PC3 Multiome_PC4</span>
<span class="co">## AAACAGCCAAGGAATC     7.914327   -1.3703933    2.8728293    0.2016397</span>
<span class="co">## AAACAGCCAGTAGGTG     9.607305   -2.7416853    2.4949054   -0.1657897</span>
<span class="co">## AAACATGCACCGGCTA     3.551898   -1.3632683   -9.1696271    0.3137612</span>
<span class="co">## AAACATGCAGGGAGCT     4.298441   -1.1046894  -11.2718261    1.2654560</span>
<span class="co">## AAACCGAAGCTGGACC    -9.870562   -0.6587611    0.3410302    0.3198895</span>
<span class="co">##                  Multiome_PC5</span>
<span class="co">## AAACAGCCAAGGAATC    0.9913363</span>
<span class="co">## AAACAGCCAGTAGGTG    1.5115966</span>
<span class="co">## AAACATGCACCGGCTA    1.6221187</span>
<span class="co">## AAACATGCAGGGAGCT    3.1555720</span>
<span class="co">## AAACCGAAGCTGGACC    4.0684566</span></code></pre>
<p>We can reduce the dimension further using non-linear approaches such
as UMAP.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stab_umap</span> <span class="op">=</span> <span class="fu">calculateUMAP</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">stab</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab_umap</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 10032     2</span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stab_umap</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.3</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dataType</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>Here we see that the RNA and Multiome cells are fairly
well-mixed.</p>
</div>
<div class="section level2">
<h2 id="data-imputation-after-stabmap">Data imputation after StabMap<a class="anchor" aria-label="anchor" href="#data-imputation-after-stabmap"></a>
</h2>
<p>Given the joint embedding, we can predict the missing ATAC values
using <code><a href="../reference/imputeEmbedding.html">imputeEmbedding()</a></code>. We provide the data list, the
joint embedding as output from <code><a href="../reference/StabMap.html">stabMap()</a></code>. We set the
Multiome cells as reference and the RNA cells as query. This is useful
for downstream visualisation or further interpretation.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">imp</span> <span class="op">=</span> <span class="fu"><a href="../reference/imputeEmbedding.html">imputeEmbedding</a></span><span class="op">(</span>
  <span class="va">assay_list</span>,
  <span class="va">stab</span>,
  reference <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">[[</span><span class="st">"Multiome"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">assay_list</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: BiocNeighbors</span></code></pre>
<pre><code><span class="co">## Warning: package 'BiocNeighbors' was built under R version 4.1.1</span></code></pre>
<pre><code><span class="co">## Loading required package: abind</span></code></pre>
<pre><code><span class="co">## only simple (non-sparse) matrix data currently supported, converting to dense matrix</span></code></pre>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">imp</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "list"</span></code></pre>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">imp</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "Multiome"</span></code></pre>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">imp</span>, <span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $Multiome</span>
<span class="co">## [1] 1740 5049</span></code></pre>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">imp</span><span class="op">[[</span><span class="st">"Multiome"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##        AAACAGCCAATCCCTT AAACAGCCAATGCGCT AAACAGCCAGTTTACG AAACAGCCATCCAGGT</span>
<span class="co">## CA6           0.0000000        0.1927288        0.0000000                0</span>
<span class="co">## CNR2          0.0000000        0.0000000        0.0000000                0</span>
<span class="co">## IFNLR1        0.0000000        0.0000000        0.0000000                0</span>
<span class="co">## RCAN3         0.7089489        1.0872543        0.9397698                0</span>
<span class="co">## ZNF683        0.0000000        0.0000000        0.0000000                0</span>
<span class="co">##        AAACATGCAAGGTCCT</span>
<span class="co">## CA6           0.7249193</span>
<span class="co">## CNR2          0.0000000</span>
<span class="co">## IFNLR1        0.0000000</span>
<span class="co">## RCAN3         0.1971583</span>
<span class="co">## ZNF683        0.0000000</span></code></pre>
</div>
<div class="section level2">
<h2 id="disjoint-mosaic-data-integration-with-stabmap">Disjoint mosaic data integration with StabMap<a class="anchor" aria-label="anchor" href="#disjoint-mosaic-data-integration-with-stabmap"></a>
</h2>
<p>StabMap is a flexible framework for mosaic data integration, and can
still integrate data even when there are pairs of datasets that share no
features at all. So long as there is a path connecting the datasets
along the mosaic data topology (and the underlying assumption that the
shared features along these paths contain information), then we can
extract meaningful joint embeddings. To demonstrate this, we will
simulate three data sources.</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataTypeDisjoint</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span>, <span class="st">"Multiome"</span>, <span class="st">"ATAC"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span>,
                                   prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>,<span class="fl">0.3</span>, <span class="fl">0.3</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
                            <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">logcounts_all</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">dataTypeDisjoint</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## dataTypeDisjoint</span>
<span class="co">##     ATAC Multiome      RNA </span>
<span class="co">##     3338     3359     3335</span></code></pre>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">assay_list_disjoint</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
  RNA <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span><span class="op">)</span>, <span class="va">dataTypeDisjoint</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RNA"</span><span class="op">)</span><span class="op">]</span>,
  Multiome <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rna"</span>, <span class="st">"atac"</span><span class="op">)</span>, <span class="va">dataTypeDisjoint</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span><span class="op">]</span>,
  ATAC <span class="op">=</span> <span class="va">logcounts_all</span><span class="op">[</span><span class="va">assayType</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"atac"</span><span class="op">)</span>, <span class="va">dataTypeDisjoint</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ATAC"</span><span class="op">)</span><span class="op">]</span>
<span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list_disjoint</span>, <span class="va">dim</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $RNA</span>
<span class="co">## [1]  952 3335</span>
<span class="co">## </span>
<span class="co">## $Multiome</span>
<span class="co">## [1] 1740 3359</span>
<span class="co">## </span>
<span class="co">## $ATAC</span>
<span class="co">## [1]  788 3338</span></code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">assay_list_disjoint</span>, <span class="va">class</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $RNA</span>
<span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span>
<span class="co">## </span>
<span class="co">## $Multiome</span>
<span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span>
<span class="co">## </span>
<span class="co">## $ATAC</span>
<span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span></code></pre>
<p>Using <code><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet()</a></code>, we note that there are no
shared features between the ATAC and RNA datasets. We might be able to
match features by extracting genomic positions and making the “central
dogma assumption”, that is, that the peaks associated with a genomic
position overlapping a gene should correspond to positive gene
expression for that gene. However, we need not make this assumption for
the data integration to be performed.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/mosaicDataUpSet.html">mosaicDataUpSet</a></span><span class="op">(</span><span class="va">assay_list_disjoint</span>, plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We can understand the <code><a href="../reference/mosaicDataTopology.html">mosaicDataTopology()</a></code> of these
datasets, which generates an <code>igraph</code> object, which can be
inspected and plotted.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mdt_disjoint</span> <span class="op">=</span> <span class="fu"><a href="../reference/mosaicDataTopology.html">mosaicDataTopology</a></span><span class="op">(</span><span class="va">assay_list_disjoint</span><span class="op">)</span>
<span class="va">mdt_disjoint</span></code></pre></div>
<pre><code><span class="co">## IGRAPH 101b86d UNW- 3 2 -- </span>
<span class="co">## + attr: name (v/c), frame.color (v/c), color (v/c), label.color (v/c),</span>
<span class="co">## | label.family (v/c), weight (e/n)</span>
<span class="co">## + edges from 101b86d (vertex names):</span>
<span class="co">## [1] RNA     --Multiome Multiome--ATAC</span></code></pre>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">mdt_disjoint</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>StabMap only requires that the mosaic data topology network be
connected, that is, that there should be a path between every pair of
nodes in the network. Since there is a path between RNA and ATAC (via
Multiome), we can proceed.</p>
<p>We now generate a common joint embedding for these data using
StabMap. Since the Multiome data contains all features, we again treat
this as the reference dataset. Since we already examined the mosaic data
topology, we set <code>plot = FALSE</code>.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stab_disjoint</span> <span class="op">=</span> <span class="fu"><a href="../reference/StabMap.html">stabMap</a></span><span class="op">(</span><span class="va">assay_list_disjoint</span>,
                        reference_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Multiome"</span><span class="op">)</span>,
                        plot <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## treating "Multiome" as reference</span></code></pre>
<pre><code><span class="co">## generating embedding for path with reference "Multiome": "Multiome"</span></code></pre>
<pre><code><span class="co">## generating embedding for path with reference "Multiome": "ATAC" -&gt; "Multiome"</span></code></pre>
<pre><code><span class="co">## generating embedding for path with reference "Multiome": "RNA" -&gt; "Multiome"</span></code></pre>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab_disjoint</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 10032    50</span></code></pre>
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stab_disjoint</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">##                  Multiome_PC1 Multiome_PC2 Multiome_PC3 Multiome_PC4</span>
<span class="co">## AAACAGCCAATCCCTT     6.859810    -1.540596    -1.843308  -0.16435493</span>
<span class="co">## AAACATGCAGCAAGTG     5.570928    -2.379867     2.911351   2.32630567</span>
<span class="co">## AAACATGCAGGGAGCT     4.550189    -1.769688   -13.014403  -0.30613633</span>
<span class="co">## AAACCAACACTAAGAA     7.888117    -1.354557     3.365346  -0.09086506</span>
<span class="co">## AAACCAACAGGATGGC     6.426812    -1.347158     1.944541  -0.12714915</span>
<span class="co">##                  Multiome_PC5</span>
<span class="co">## AAACAGCCAATCCCTT   -1.5442658</span>
<span class="co">## AAACATGCAGCAAGTG   -1.7983382</span>
<span class="co">## AAACATGCAGGGAGCT    1.9258897</span>
<span class="co">## AAACCAACACTAAGAA    0.6042899</span>
<span class="co">## AAACCAACAGGATGGC   -1.8859162</span></code></pre>
<p>We can reduce the dimension further using non-linear approaches such
as UMAP.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stab_disjoint_umap</span> <span class="op">=</span> <span class="fu">calculateUMAP</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">stab_disjoint</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">stab_disjoint_umap</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] 10032     2</span></code></pre>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stab_disjoint_umap</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.3</span>,
     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">dataTypeDisjoint</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab_disjoint</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>Here we see that the RNA, ATAC and Multiome cells are fairly
well-mixed.</p>
<p>Colouring the cells by their original cell type, we can also see that
the mosaic data integration is meaningful.</p>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cellType</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">$</span><span class="va">celltype</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stab_disjoint_umap</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.3</span>,
     col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">cellType</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stab_disjoint</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="stabMap_PBMC_Multiome_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.0 (2021-05-18)</span>
<span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">## Running under: macOS Catalina 10.15.7</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.dylib</span>
<span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] abind_1.4-5                 BiocNeighbors_1.12.0       </span>
<span class="co">##  [3] scater_1.22.0               ggplot2_3.3.6              </span>
<span class="co">##  [5] UpSetR_1.4.0                scran_1.22.1               </span>
<span class="co">##  [7] scuttle_1.4.0               SingleCellExperiment_1.16.0</span>
<span class="co">##  [9] SingleCellMultiModal_1.6.0  MultiAssayExperiment_1.20.0</span>
<span class="co">## [11] SummarizedExperiment_1.24.0 Biobase_2.54.0             </span>
<span class="co">## [13] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1        </span>
<span class="co">## [15] IRanges_2.28.0              S4Vectors_0.32.3           </span>
<span class="co">## [17] BiocGenerics_0.40.0         MatrixGenerics_1.6.0       </span>
<span class="co">## [19] matrixStats_0.61.0          StabMap_0.1.2              </span>
<span class="co">## [21] igraph_1.2.11               BiocStyle_2.22.0           </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] AnnotationHub_3.2.2           BiocFileCache_2.2.1          </span>
<span class="co">##   [3] systemfonts_1.0.4             plyr_1.8.7                   </span>
<span class="co">##   [5] BiocParallel_1.28.3           digest_0.6.29                </span>
<span class="co">##   [7] htmltools_0.5.2               viridis_0.6.2                </span>
<span class="co">##   [9] magick_2.7.3                  fansi_1.0.3                  </span>
<span class="co">##  [11] magrittr_2.0.3                memoise_2.0.1                </span>
<span class="co">##  [13] ScaledMatrix_1.2.0            SpatialExperiment_1.4.0      </span>
<span class="co">##  [15] cluster_2.1.2                 limma_3.50.1                 </span>
<span class="co">##  [17] Biostrings_2.62.0             R.utils_2.11.0               </span>
<span class="co">##  [19] pkgdown_2.0.2                 colorspace_2.0-3             </span>
<span class="co">##  [21] ggrepel_0.9.1                 blob_1.2.2                   </span>
<span class="co">##  [23] rappdirs_0.3.3                textshaping_0.3.6            </span>
<span class="co">##  [25] xfun_0.29                     dplyr_1.0.9                  </span>
<span class="co">##  [27] crayon_1.5.1                  RCurl_1.98-1.6               </span>
<span class="co">##  [29] jsonlite_1.8.0                glue_1.6.2                   </span>
<span class="co">##  [31] gtable_0.3.0                  zlibbioc_1.40.0              </span>
<span class="co">##  [33] XVector_0.34.0                DelayedArray_0.20.0          </span>
<span class="co">##  [35] BiocSingular_1.10.0           DropletUtils_1.14.2          </span>
<span class="co">##  [37] Rhdf5lib_1.16.0               HDF5Array_1.22.1             </span>
<span class="co">##  [39] scales_1.2.0                  DBI_1.1.2                    </span>
<span class="co">##  [41] edgeR_3.36.0                  Rcpp_1.0.8.3                 </span>
<span class="co">##  [43] viridisLite_0.4.0             xtable_1.8-4                 </span>
<span class="co">##  [45] dqrng_0.3.0                   bit_4.0.4                    </span>
<span class="co">##  [47] rsvd_1.0.5                    metapod_1.2.0                </span>
<span class="co">##  [49] httr_1.4.2                    ellipsis_0.3.2               </span>
<span class="co">##  [51] farver_2.1.0                  pkgconfig_2.0.3              </span>
<span class="co">##  [53] R.methodsS3_1.8.1             uwot_0.1.11                  </span>
<span class="co">##  [55] sass_0.4.0                    dbplyr_2.1.1                 </span>
<span class="co">##  [57] locfit_1.5-9.4                utf8_1.2.2                   </span>
<span class="co">##  [59] labeling_0.4.2                tidyselect_1.1.2             </span>
<span class="co">##  [61] rlang_1.0.2                   later_1.3.0                  </span>
<span class="co">##  [63] AnnotationDbi_1.56.2          munsell_0.5.0                </span>
<span class="co">##  [65] BiocVersion_3.14.0            tools_4.1.0                  </span>
<span class="co">##  [67] cachem_1.0.6                  cli_3.3.0                    </span>
<span class="co">##  [69] generics_0.1.2                RSQLite_2.2.10               </span>
<span class="co">##  [71] ExperimentHub_2.2.1           evaluate_0.15                </span>
<span class="co">##  [73] stringr_1.4.0                 fastmap_1.1.0                </span>
<span class="co">##  [75] yaml_2.3.5                    ragg_1.2.2                   </span>
<span class="co">##  [77] knitr_1.37                    bit64_4.0.5                  </span>
<span class="co">##  [79] fs_1.5.2                      purrr_0.3.4                  </span>
<span class="co">##  [81] KEGGREST_1.34.0               sparseMatrixStats_1.6.0      </span>
<span class="co">##  [83] mime_0.12                     R.oo_1.24.0                  </span>
<span class="co">##  [85] compiler_4.1.0                rstudioapi_0.13              </span>
<span class="co">##  [87] beeswarm_0.4.0                filelock_1.0.2               </span>
<span class="co">##  [89] curl_4.3.2                    png_0.1-7                    </span>
<span class="co">##  [91] interactiveDisplayBase_1.32.0 tibble_3.1.7                 </span>
<span class="co">##  [93] statmod_1.4.36                bslib_0.3.1                  </span>
<span class="co">##  [95] stringi_1.7.6                 highr_0.9                    </span>
<span class="co">##  [97] RSpectra_0.16-1               desc_1.4.0                   </span>
<span class="co">##  [99] lattice_0.20-45               bluster_1.4.0                </span>
<span class="co">## [101] Matrix_1.4-0                  vctrs_0.4.1                  </span>
<span class="co">## [103] pillar_1.7.0                  lifecycle_1.0.1              </span>
<span class="co">## [105] rhdf5filters_1.6.0            BiocManager_1.30.16          </span>
<span class="co">## [107] jquerylib_0.1.4               RcppAnnoy_0.0.19             </span>
<span class="co">## [109] bitops_1.0-7                  irlba_2.3.5                  </span>
<span class="co">## [111] httpuv_1.6.5                  R6_2.5.1                     </span>
<span class="co">## [113] bookdown_0.24                 promises_1.2.0.1             </span>
<span class="co">## [115] gridExtra_2.3                 vipor_0.4.5                  </span>
<span class="co">## [117] codetools_0.2-18              MASS_7.3-55                  </span>
<span class="co">## [119] assertthat_0.2.1              rhdf5_2.38.0                 </span>
<span class="co">## [121] rprojroot_2.0.2               rjson_0.2.21                 </span>
<span class="co">## [123] withr_2.5.0                   GenomeInfoDbData_1.2.7       </span>
<span class="co">## [125] parallel_4.1.0                grid_4.1.0                   </span>
<span class="co">## [127] beachmat_2.10.0               rmarkdown_2.12               </span>
<span class="co">## [129] DelayedMatrixStats_1.16.0     shiny_1.7.1                  </span>
<span class="co">## [131] ggbeeswarm_0.6.0</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Shila Ghazanfar.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
